lightning/src/routing/router.rs

```diff
@@ -876,40 +877,115 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
-					// If this hop connects to a node with which we have a direct channel,
-					// ignore the network graph and, if the last hop was added, add our
-					// direct channel to the candidate set.
-					//
-					// Note that we *must* check if the last hop was added as `add_entry`
-					// always assumes that the third argument is a node to which we have a
-					// path.
-					if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&hop.src_node_id) {
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, 0, path_value_msat, 0);
+				// We start building the path from reverse, i.e., from payee
+				// to the first RouteHintHop in the path.
+				let route_it = route.0.iter().rev();
+				let mut prev_route_it = route.0.iter().rev();
+				let mut aggregate_next_hops_fee_msat: u64 = 0;
+				let mut aggregate_next_hops_path_htlc_minimum_msat: u64 = 0;
+				let mut prev_hop;
+
+				for (idx, hop) in route_it.enumerate(){
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
+					let hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+						Some(htlc_minimum_msat) => htlc_minimum_msat,
+						None => 0
+					};
+
+					let directional_info = DummyDirectionalChannelInfo {
+						cltv_expiry_delta: hop.cltv_expiry_delta as u32,
+						htlc_minimum_msat: hop_htlc_minimum_msat,
+						htlc_maximum_msat: hop.htlc_maximum_msat,
+						fees: hop.fees,
+					};
+
+					prev_hop = match idx {
+						0 => &payee,
+						_ => match prev_route_it.next() {
+							Some(x) => &x.src_node_id,
+							None => &payee,
+						}
+					};
+					let reqd_channel_cap = Some(match idx {
+						0 => (final_value_msat + 999),
+						_ => match aggregate_next_hops_fee_msat.checked_add(999) {
+							Some(cap) => cap,
+							None => u64::max_value()
+						},
+					} / 1000); // converting from msat
+
+
+					// We assume that the recipient only included route hints for routes which had
+					// sufficient value to route `final_value_msat`. Note that in the case of "0-value"
+					// invoices where the invoice does not specify value this may not be the case, but
+					// better to include the hints than not.
+					if add_entry!(hop.short_channel_id, hop.src_node_id, prev_hop, directional_info, reqd_channel_cap, &empty_channel_features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat) {
+
+						// If the index is 0, then its the last hop in the RoutHint. The
+						// aggregate_next_hops_fee_msat for this hop was 0 but from here on out
+						// it needs to be increased beyond the value to be routed to account
+						// for the fee charged at hops and value to send from payer to payee.
+						if idx == 0 {
+							aggregate_next_hops_fee_msat = match aggregate_next_hops_fee_msat.checked_add(final_value_msat) {
```

Comment:
Hmm, I may have been confusing in my previous review - `aggregate_next_hops_fee_msat` should not *include* `final_value_msat`, it should always be only the fee. However, the calls to `compute_fee` should include `final_value_msat` because the fees are always calculated against the total value at a hop, not against just the fee portion.

Reply:
Okay. I think I understand that. Adding that modification. I noticed it now that the test ```last_hops_multi_hint_with_disabled_channels_test``` fails with this modification even though it's supposed to find a path to  ```node6`` through channels ```5``` and ```8```. It errors out with ```{ err: "Failed to find a path to the given destination", action: IgnoreError }', lightning/src/routing/router.rs:2468:228```